src/transformers/trainer.py

```diff
@@ -609,7 +616,9 @@ def create_optimizer_and_scheduler(self, num_training_steps: int):
             self.lr_scheduler = get_scheduler(
                 self.args.lr_scheduler_type,
                 self.optimizer,
-                num_warmup_steps=self.args.warmup_steps,
+                num_warmup_steps=self.args.warmup_steps
+                if self.args.warmup_steps > 0
+                else math.ceil(num_training_steps * self.args.warmup_ratio),
```

Comment:
Small nit: this is not super readable so it would be great to define before this function
```
wamup_steps = self.args.warmup_steps if self.args.warmup_steps > 0 else math.ceil(num_training_steps * self.args.warmup_ratio)
```
(which will be restyled since it's a bit long so you will need to run make style) and then use `warmup_steps` here.